### DISCUSSION TITLE
Nullable nested data fails validation when nested data attributes are required

### DISCUSSION THREAD

**Discussion by ominixamf** [spatie/laravel-data#55] [Q&A] [answered]
Not sure if this behavior is expected, but here we go:

I have two data objects (simplified below)

```
class CompanyCnpjData extends Data
{
    public function __construct(
        public string $cnpj,

        /** @var \App\Data\ActivityData */
        public ?ActivityData $mainActivity,

        /** @var \App\Data\ActivityData[] */
        public ?DataCollection $otherActivities
    ) {
    }
}
```

and

```
class ActivityData extends Data
{
    public function __construct(
        #[Regex('/^[0-9]+$/')]
        public string $code,
        public string $description,
    ) {
    }
}
```

The idea is that both $mainActivity and $otherActivities in the first Data object can be nullable.

However, when `$mainActivity` is null, validation fails with errors on `mainActivity.code` and `mainActivity.description` as required.

Here are the rules I see the package is generating for these attributes (debug from within `DataValidationRulesResolver`):

```
[mainActivity] => Array
        (
            [0] => nullable
            [1] => array
        )

    [mainActivity.code] => Array
        (
            [0] => string
            [1] => regex:/^[0-9]+$/
            [2] => required
        )

    [mainActivity.description] => Array
        (
            [0] => string
            [1] => required
        )
```

The same doesn't happen for the `DataCollection $otherActivities`, i.e., it is indeed nullable

Am I missing any configuration on the `CompanyCnpjData nested `$mainActivity`? Is this laravel validator default behavior or maybe a case not handled by the package?

**Comment by rubenvanassche** [comment, marked as answer]
Hi @ominixamf,

This was indeed a bug, it should be fixed with version 1.3.0!

**Reply by leonampuzlplace** [reply]
same problem here. Already some solution?
im using "spatie/laravel-data": "^1.4" version

Sorry for my english, im a brazilian guy.
thx

**Reply by thewebartisan7** [reply]
@rubenvanassche sound like my same issue I had..

**Reply by raveren** [reply]
This ain't solved, is there an open issue for this?

**Comment by gitmiro** [comment]
Hey guys

First of all, this lib is awesome! Really!

I just came across an issue which confuses me a bit, as it was already solved and yet I have the same error with v2.0.12.

```php
class UpdateUser extends Data
{
    public function __construct(
        public ?SetUserProfile $setUserProfile,
        public ?SetUserRoles $setUserRoles
    ) {
    }
}



class SetUserProfile extends Data
{
    public function __construct(
        #[Required] public string $name
    ) {
    }
}


```


Running this through a request with the following payload: 


```php

$payload = [
    'setUserRoles' => ['some', roles]
]

```


Results in a validation-error: The set user profile.name field is required.


Is this the intended behaviour? If not, how can I stop the validation on nested Data-objects if none has been provided?

Cheers!

**Reply by thewebartisan7** [reply]
It seem like issue that I am facing.
The only solution that worked for me was to add without validation attribute:

#[WithoutValidation]

As it seem that from request it validate recursively all collection, and this seem as designed.

**Reply by gitmiro** [reply]
well a workaround which works for me is the `#[RequiredWith('setUserProfile)]` Attribute. Not beautiful but it does the job

**Reply by francoism90** [reply]
@gitmiro Could you try https://github.com/spatie/laravel-data/pull/226? I think it's the same issue.

**Reply by gitmiro** [reply]
@francoism90 I actually tried your approach but it did not work out for me. The validation was still being triggered

**Reply by francoism90** [reply]
Could you try https://laravel.com/docs/9.x/validation#rule-exclude-without? It seems to work for me.

**Reply by gitmiro** [reply]
@francoism90 thanks for the hint! Indeed another solution. But I still have to place it over each property (with the parent Class name) as in my example with RequiredWith. But I think I will extend my RuleInferrer to handle that case as well

**Comment by gitmiro** [comment]
Just if anybody is interested, the ` #[RequiredWith('setUserProfile)]` was too cumbersome for me in the end. I ended up using e custom RuleInferrer with my own Attribute Annotation:

```php
class RequiredWithThisRuleInferrer implements RuleInferrer
{
    public function handle(
        DataProperty $property,
        RulesCollection $rules
    ): RulesCollection {
        $property->attributes
            ->filter(fn(object $attribute) => $attribute instanceof RequiredWithThis)
            ->each(fn(RequiredWithThis $rule) => $rule->setClass($property->className));

        return $rules;
    }
}

```

The Attribute which will be used for validation:

```php
#[Attribute(Attribute::TARGET_PROPERTY)]
class RequiredWithThis extends StringValidationAttribute implements RequiringRule
{
    protected string $class;

    public static function keyword(): string
    {
        return 'required_with';
    }

    public function setClass(string $class)
    {
        $name = collect(explode('\\', $class))->last();
        $this->class = Str::camel($name);
    }

    public function parameters(): array
    {
        return [$this->normalizeValue($this->class)];
    }
}

```

And finally, how I use it:

```php
class SetUserProfile extends Data
{
    public function __construct(
       #[RequiredWithThis] public string $firstname,
       #[RequiredWithThis] public string $lastname
    ) {
    }
}
```

Rendererd rule:

```php
 "setUserProfile.firstname" => array:3 [
    0 => "nullable"
    1 => "string"
    2 => "required_with:setUserProfile"
  ]
```


Again, maybe not the best way. But at least the solution is not interfering with the core of the package.

Any better suggestions are welcome!

**Comment by rubenvanassche** [comment]
The problem is kinda similar to the issue @francoism90 made, though not using 'sometimes' but 'nullable'. I've asked the Laravel developers what the best solution would be: https://github.com/laravel/framework/issues/44586

**Reply by raveren** [reply]
I'm sorry to ask, but is this actively worked on? 

The test case is pretty clearly defined, 'just' needs some clever solution to not validate subkeys of unset optional objects..

**Reply by francoism90** [reply]
@raveren Solutions/workarounds are already posted: `required_with`, `exclude_without`, `sometimes`

**Comment by raveren** [comment]
We've got progress! Hallelujah! Active github issue:

https://github.com/spatie/laravel-data/issues/309
